### DISCUSSION TITLE
auto proxySet/proxyMap in proxy, and auto new Set/ new Map in snapshot

### DISCUSSION THREAD

**Discussion by magicdawn** [pmndrs/valtio#915] [Ideas]
background, when persist data is needed, serialization / deserialization is need. 
the [superjson](https://github.com/blitz-js/superjson) supports Map/Set serialization / deserialization

when use set/map with valtio, things need to be done

1. load storage -> superjson.parse(may contains Map/Set) -> transform map/set to proxySet/proxyMap -> valtio.proxy(transformed)
2. when state change -> valtio.snapshot(state) -> transform snapshot-of-proxySet/proxyMap to real Set/Map -> superjson.stringify -> save to storage

AS u can see, it would be nice for `proxy()` support real `Set/Map`: auto transform to proxySet/proxyMap

**Comment by dai-shi** [comment]
Thanks for opening up the discussion.

Valtio's design decision is that it doesn't support native Set/Map. So, proxySet/Map is implemented as an emulation.
Now, your suggestion sounds reasonable. Currently, customizing set trap of a Valtio proxy isn't supported. And, only the solution is a low-level function `unstable_buildProxyFunction`. 
 https://github.com/pmndrs/valtio/discussions/909#discussioncomment-9690194

We would like to collect more feedback on the low-level function and re-design a new/better/extensible API for the future version of Valtio.

**Comment by dai-shi** [comment]
Auto proxySet/Map might be technically possible with new internals exposed.

Also, I opened up a new discussion: #943
